### PR TITLE
MRG: TFMxNE using BCD and active set

### PR DESCRIPTION
**Pull request by yousrabk**
Taking over @joewalter's PR (closes #2203)

This is removing the different solvers and keeps only BCD with active set approach, which is having "same" source estimate much faster.

### CODE REVIEW

File: mne/inverse_sparse/mxne_inverse.py [hunk 1]
```diff
@@ -110,6 +110,10 @@ def _compute_residual(forward, evoked, X, active_set, info):
     residual = evoked.copy()
     residual = pick_channels_evoked(residual, include=info['ch_names'])
     r_tmp = residual.copy()
+
+    print(forward['sol']['data'][sel, :].shape)
+    print(active_set.shape, X.shape)
```
Comment (Eric89GXL) [inline mne/inverse_sparse/mxne_inverse.py]: I don't think these are needed, they look like debugging lines

File: mne/inverse_sparse/mxne_inverse.py [hunk 2]
```diff
@@ -433,10 +437,12 @@ def tf_mixed_norm(evoked, forward, noise_cov, alpha_space, alpha_time,
...
         It can be one float or float if the values are different for left
         and right window length.
+    log_objective : bool
+        If True, the cost function is evaluated in every iteration.
     return_residual : bool
         If True, the residual is returned as an Evoked instance.
-    verbose : bool, str, int, or None
-        If not None, override default verbose level (see mne.verbose).
+    verbose: bool
+        Verbose output or not.
```
Comment (Eric89GXL) [inline mne/inverse_sparse/mxne_inverse.py]: -1, this goes against how `verbose` is used and described everywhere else in the package

File: mne/inverse_sparse/mxne_optim.py [hunk 3]
```diff
@@ -670,10 +673,361 @@ def __call__(self, z):
...
+    if Z.shape[0]:
+        Z2 = Z.reshape(*shape)
+        l21_norm = np.sqrt(stft_norm2(Z2).reshape(-1, n_orient).sum(axis=1))
+        l21_norm = l21_norm.sum()
+    else:
+        l21_norm = 0.
+    return l21_norm
+
+
+@verbose
```
Comment (Eric89GXL) [inline mne/inverse_sparse/mxne_optim.py]: no reason to have the `verbose` decorator on a function with no `verbose` parameter, remove

File: mne/inverse_sparse/mxne_optim.py [hunk 4]
```diff
@@ -670,10 +673,361 @@ def __call__(self, z):
 
 
 @verbose
```
Comment (Eric89GXL) [inline mne/inverse_sparse/mxne_optim.py]: same here

File: mne/inverse_sparse/tests/test_mxne_optim.py [hunk 5]
```diff
@@ -104,15 +104,15 @@ def test_l21_mxne():
 
 def test_tf_mxne():
     """Test convergence of TF-MxNE solver"""
-    alpha_space = 10.
-    alpha_time = 5.
+    alpha_space = 40.
+    alpha_time = 0.1
```
Comment (Eric89GXL) [inline mne/inverse_sparse/tests/test_mxne_optim.py]: These constants should be renamed to reflect their new names. Also, there are a number of places these constants have been changed. This sounds backward-incompatible to me. Or is it just to improve the quality of the results over what we had before?

**Comment by agramfort** [comment]
@yousrabk see the diff and make sure you avoid API changes. We just change the solver here

File: mne/inverse_sparse/tests/test_mxne_optim.py (hunk 5 above)
Comment (yousrabk) [inline mne/inverse_sparse/tests/test_mxne_optim.py]: Actually it's not only the name that has been changed. May be the doc for alpha and rho are misleading, should we change that @agramfort?
@Eric89GXL they are kind of different parameters: `alpha_space = alpha * (1 - rho)` and `alpha_time = alpha * rho`

**Comment by yousrabk** [comment]
@agramfort do you mean I should keep the alpha_space and alpha time both as before? (I remove the alpha/rho?)

**Comment by agramfort** [comment]
yes keep the alpha_space and alpha time both as before

File: examples/inverse/plot_time_frequency_mixed_norm_inverse.py [hunk 6]
```diff
@@ -72,7 +72,7 @@
 alpha_space = 50.  # spatial regularization parameter
 # alpha_time parameter promotes temporal smoothness
 # (0 means no temporal regularization)
-alpha_time = 1.  # temporal regularization parameter
+alpha_time = 10.  # temporal regularization parameter
```
Comment (agramfort) [inline examples/inverse/plot_time_frequency_mixed_norm_inverse.py]: why did you change this?

File: mne/inverse_sparse/mxne_inverse.py [hunk 7]
```diff
@@ -435,6 +436,8 @@ def tf_mixed_norm(evoked, forward, noise_cov, alpha_space, alpha_time,
         and right window length.
     return_residual : bool
         If True, the residual is returned as an Evoked instance.
+    log_objective : bool
+        If True, the cost function is evaluated in every iteration.
```
Comment (agramfort) [inline mne/inverse_sparse/mxne_inverse.py]: do we think we need this param?

File: mne/inverse_sparse/mxne_optim.py [hunk 8]
```diff
@@ -669,11 +671,294 @@ def __call__(self, z):
...
+    alpha_time : float in [0, 100]
+        Regularization parameter for temporal sparsity. It set to 0,
+        no temporal regularization is applied. It this case, TF-MxNE is
+        equivalent to MxNE with L21 norm.
+    lipschitz_constant : float
+        The lipschitz constant of the spatio temporal linear operator.
+    phi : instance of _Phi
+        The forward STFT operator.
+    phiT : instance of _PhiT
+        The inverse STFT operator.
```
Comment (agramfort) [inline mne/inverse_sparse/mxne_optim.py]: The transpose of the TF operator.

File: mne/inverse_sparse/mxne_optim.py [hunk 9]
```diff
@@ -669,11 +671,294 @@ def __call__(self, z):
...
+        Regularization parameter for spatial sparsity. If larger than 100,
+        then no source will be active.
+    alpha_time : float in [0, 100]
+        Regularization parameter for temporal sparsity. It set to 0,
+        no temporal regularization is applied. It this case, TF-MxNE is
+        equivalent to MxNE with L21 norm.
+    lipschitz_constant : float
+        The lipschitz constant of the spatio temporal linear operator.
+    phi : instance of _Phi
+        The forward STFT operator.
```
Comment (agramfort) [inline mne/inverse_sparse/mxne_optim.py]: The TF operator.

File: mne/inverse_sparse/mxne_inverse.py [hunk 10]
```diff
@@ -470,23 +480,10 @@ def tf_mixed_norm(evoked, forward, noise_cov, alpha_space, alpha_time,
     logger.info('Whitening data matrix.')
     M = np.dot(whitener, M)
 
-    # Scaling to make setting of alpha easy
-    n_dip_per_pos = 1 if is_fixed_orient(forward) else 3
-    alpha_max = norm_l2inf(np.dot(gain.T, M), n_dip_per_pos, copy=False)
-    alpha_max *= 0.01
-    gain /= alpha_max
-    source_weighting /= alpha_max
```
Comment (agramfort) [inline mne/inverse_sparse/mxne_inverse.py]: why removing this?

File: mne/inverse_sparse/mxne_optim.py [hunk 11]
```diff
@@ -736,98 +1018,32 @@ def tf_mixed_norm_solver(M, G, alpha_space, alpha_time, wsize=64, tstep=4,
...
+            lc[j] = linalg.norm(np.dot(G_tmp.T, G_tmp), ord=2)
+
+    logger.info("Using block coordinate descent and active set approach")
+    X, Z, active_set, E = _tf_mixed_norm_solver_bcd_active_set(
+        M, G, alpha_space, alpha_time, lc, phi, phiT, Z_init=None,
+        wsize=wsize, tstep=tstep, n_orient=n_orient, maxit=maxit, tol=tol,
+        log_objective=log_objective, verbose=None)
+
+    if debias:
+        if active_set.sum() > 0:
```
Comment (agramfort) [inline mne/inverse_sparse/mxne_optim.py]: one if is enough with an "and"

File: mne/inverse_sparse/tests/test_mxne_inverse.py [hunk 12]
```diff
@@ -101,7 +101,7 @@ def test_mxne_inverse():
 
     # Do with TF-MxNE for test memory savings
     alpha_space = 60.  # spatial regularization parameter
-    alpha_time = 1.  # temporal regularization parameter
+    alpha_time = 10.  # temporal regularization parameter
```
Comment (agramfort) [inline mne/inverse_sparse/tests/test_mxne_inverse.py]: why?

**Comment by joewalter** [comment]
In my opinion, being able to compute an upper bound on the regularization parameters and connecting the two parameters was one of the best things in the pull request, so I don't understand why it is eliminated. With the current code the space and time regularization parameters don't fit, i.e. even with both parameters < 100 the active set size can be zero, which is not what I expect as a user (in my opinion).

File: examples/inverse/plot_time_frequency_mixed_norm_inverse.py (hunk 6 above)
Comment (yousrabk) [inline examples/inverse/plot_time_frequency_mixed_norm_inverse.py]: I was just trying different parameters, missed that on the diff.

**Comment by yousrabk** [comment]
@joewalter you mean why don't we keep alpha/rho?

**Comment by joewalter** [comment]
@yousrabk yes, not the name of the variables but the way they are defined and normalized

**Comment by agramfort** [comment]
the alpha_space and alpha_time are the parametrization we used in the
neuroimage paper

**Comment by yousrabk** [comment]
The aim was to not change the API used before

**Comment by agramfort** [comment]
let's have an API change in another PR

**Comment by joewalter** [comment]
Ok sorry for the noise, just my feeling.

File: mne/inverse_sparse/mxne_inverse.py (hunk 7 above)
Comment (yousrabk) [inline mne/inverse_sparse/mxne_inverse.py]: We can remove it and compute the cost function all the time

Comment (agramfort) [inline mne/inverse_sparse/mxne_inverse.py]: was it like that before?

Comment (yousrabk) [inline mne/inverse_sparse/mxne_inverse.py]: Yes the log_objective was here even before

Comment (agramfort) [inline mne/inverse_sparse/mxne_inverse.py]: ok so keep it

**Comment by yousrabk** [comment]
Updated

File: mne/inverse_sparse/mxne_inverse.py [hunk 13]
```diff
@@ -471,22 +481,21 @@ def tf_mixed_norm(evoked, forward, noise_cov, alpha_space, alpha_time,
...
-                                            alpha_space, alpha_time,
-                                            wsize=wsize, tstep=tstep,
-                                            maxit=maxit, tol=tol,
-                                            verbose=verbose,
-                                            n_orient=n_dip_per_pos,
-                                            debias=debias)
+    X, active_set, E = tf_mixed_norm_solver(
+        M, gain, alpha_space, alpha_time, wsize=wsize, tstep=tstep,
+        maxit=maxit, tol=tol, verbose=verbose, n_orient=n_dip_per_pos,
+        log_objective=log_objective, debias=debias)
```
Comment (agramfort) [inline mne/inverse_sparse/mxne_inverse.py]: you log the objective but what do you do with E? you don't return it right?

File: mne/inverse_sparse/mxne_inverse.py [hunk 14]
```diff
@@ -446,15 +449,22 @@ def tf_mixed_norm(evoked, forward, noise_cov, alpha_space, alpha_time,
         The residual a.k.a. data not explained by the sources.
         Only returned if return_residual is True.
     """
-    _check_reference(evoked)
```
Comment (agramfort) [inline mne/inverse_sparse/mxne_inverse.py]: can you put it back?

File: mne/inverse_sparse/mxne_inverse.py [hunk 15]
```diff
@@ -446,15 +449,22 @@ def tf_mixed_norm(evoked, forward, noise_cov, alpha_space, alpha_time,
...
+                        ' Got alpha_space = %f' % alpha_space)
+
+    if (alpha_time < 0.) or (alpha_time > 100.):
+        raise Exception('alpha_time must be in range [0, 100].'
+                        ' Got alpha_time = %f' % alpha_time)
+
     # put the forward solution in fixed orientation if it's not already
     if loose is None and not is_fixed_orient(forward):
         forward = deepcopy(forward)
         _to_fixed_ori(forward)
```
Comment (agramfort) [inline mne/inverse_sparse/mxne_inverse.py]: insert empty line

File: mne/inverse_sparse/mxne_optim.py [hunk 16]
```diff
@@ -736,95 +1016,28 @@ def tf_mixed_norm_solver(M, G, alpha_space, alpha_time, wsize=64, tstep=4,
         is False, it will be empty.
     """
     n_sensors, n_times = M.shape
-    n_dipoles = G.shape[1]
+    n_sensors, n_sources = G.shape
+    n_positions = n_sources // n_orient
 
     n_step = int(ceil(n_times / float(tstep)))
-    n_freq = wsize // 2 + 1
+    n_freq = wsize / 2 + 1
```
Comment (agramfort) [inline mne/inverse_sparse/mxne_optim.py]: put the // back

**Comment by agramfort** [comment]
do you get exactly the same results with this branch as with master? can you at least test on 2 datasets?

File: mne/inverse_sparse/mxne_inverse.py (hunk 13 above)
Comment (yousrabk) [inline mne/inverse_sparse/mxne_inverse.py]: True, it's only returned in the solver

File: mne/inverse_sparse/mxne_inverse.py (hunk 14 above)
Comment (yousrabk) [inline mne/inverse_sparse/mxne_inverse.py]: yes

**Comment by yousrabk** [comment]
Yes it's exactly the same results compared to master. Ok I can add another test

File: mne/inverse_sparse/mxne_inverse.py (hunk 13 above)
Comment (joewalter) [inline mne/inverse_sparse/mxne_inverse.py]: That's also someething we should keep in mind for the API change pull request. If we compute the cost function, then why shouldn't we return it. While implementing these funtions I had always versions giving back time and cost function per iteration to check convergence and speed. The problem that I see is that in this case we have two parameters changing the output of the main function, i.e. return_residual and log_objective. So we have to be more careful while calling the function. Maybe theres another more secure idea, e.g. always returing E and residual which are None if return_residual=None and log_objective=None.

**Comment by joewalter** [comment]
Cool that it gives exactly the same result. I expected a small diff due to the issues of convergence you described for the FISTA in master.

**Comment by agramfort** [comment]
don't add a test as the old solver will disappear

File: mne/inverse_sparse/mxne_inverse.py (hunk 13 above)
Comment (agramfort) [inline mne/inverse_sparse/mxne_inverse.py]: >
> That's also someething we should keep in mind for the API change pull request. If we compute the cost function, then why shouldn't we return it. While implementing these funtions I had always versions giving back time and cost function per iteration to check convergence and speed. The problem that I see is that in this case we have two parameters changing the output of the main function, i.e. return_residual and log_objective. So we have to be more careful while calling the function. Maybe theres another more secure idea, e.g. always returing E and residual which are None if return_residual=None and log_objective=None.
>
>
ok but we should not mix debug statements and testing on devel with what
the user should have access to. The code by default should not compute
anything which is just used for debug or profiling.

**Comment by joewalter** [comment]
I am not sure if we should eliminate FISTA completely as FISTA or an improved version of it can be better in a parallel setting. Isn't it enough to change the default ('auto') to BCD and keep FISTA in the code (it's the solver, which we described in the Neuroimage paper, so my feeling is to keep it somewhere).

**Comment by agramfort** [comment]
can FISTA be faster? I don't want to maintain code if there is no reason to
use it.

**Comment by yousrabk** [comment]
Well, I was not completely right, I tested it on visual and audit (left/right). The amplitude are different :/
I think I'm missing something on the computation of alpha_max

**Comment by agramfort** [comment]
travis is not happy. Did you rebase on clean master?

**Comment by yousrabk** [comment]
Yes I actually deleted my master and pulled it back. So it must be clean and up to date.

File: mne/inverse_sparse/mxne_inverse.py [hunk 17]
```diff
@@ -364,8 +365,8 @@ def _window_evoked(evoked, size):
 def tf_mixed_norm(evoked, forward, noise_cov, alpha_space, alpha_time,
                   loose=0.2, depth=0.8, maxit=3000, tol=1e-4,
                   weights=None, weights_min=None, pca=True, debias=True,
-                  wsize=64, tstep=4, window=0.02,
-                  return_residual=False, verbose=None):
+                  wsize=64, tstep=4, window=0.02, return_residual=False,
+                  log_objective=True, verbose=None):
```
Comment (agramfort) [inline mne/inverse_sparse/mxne_inverse.py]: why this new log_objective param ?

File: mne/inverse_sparse/mxne_inverse.py (hunk 7 above)
Comment (agramfort) [inline mne/inverse_sparse/mxne_inverse.py]: remove it. As it is only used in optim solver.

File: mne/inverse_sparse/mxne_inverse.py (hunk 13 above)
Comment (agramfort) [inline mne/inverse_sparse/mxne_inverse.py]: set log_objective to False here. As it's not useful for end user.

File: mne/inverse_sparse/mxne_inverse.py (hunk 17 above)
Comment (yousrabk) [inline mne/inverse_sparse/mxne_inverse.py]: Yes, no need of this param here.

Comment (agramfort) [inline mne/inverse_sparse/mxne_inverse.py]: let me know when you're done cleaning things up. thx

Comment (yousrabk) [inline mne/inverse_sparse/mxne_inverse.py]: Ok removed. And even Travis is happy now

**Comment by agramfort** [comment]
merge by rebase after updating what's new.

thanks @yousrabk and @joewalter !
